src/traction_modem/MemorySpace.hxx

```diff
@@ -0,0 +1,405 @@
+            return NONE;
+        };
+
+        MemorySpace *parent_; ///< parent object
+    } timer_;
+
+    Notifiable *done_; ///< Notifiable for the memory operation to continue.
+    uint8_t *rdData_; ///< read data pointer
+    size_t size_; ///< requested size
+    State state_; ///< current request state
+    errorcode_t error_; ///< unsupported space
+
+    /// Allow access from child timer object.
+    friend class timeout;
+};
+
+/// Memory space for DCC CVs.
+class CvSpace : public MemorySpace
+{
+public:
+    /// Memory space number for CVs
+    static constexpr uint8_t SPACE_ID = openlcb::MemoryConfigDefs::SPACE_DCC_CV;
+
+    /// Constructor.
+    /// @param service Service instance to bind this flow to
+    /// @param tx_flow reference to the transmit flow
+    /// @param rx_flow reference to the receive flow
+    CvSpace(
+        Service *service, TxFlowInterface *tx_flow, RxFlowInterface *rx_flow)
+        : MemorySpace(service, tx_flow, rx_flow)
+    {
+    }
+
+private:
+    /// Test if the memory space is read only.
+    /// @return false
+    bool read_only() override
+    {
+        return false;
+    }
+
+    /// Get the largest supported address.
+    /// @return 1023 (max CV address)
+    address_t max_address() override
+    {
+        return 1023;
+    };
+
+    /// Get the space ID that will be used over the modem interface.
+    /// @return space id
+    uint8_t get_space_id() override
+    {
+        return SPACE_ID;
+    }
+};
+
+/// Memory space for firmware updates.
+class FirmwareSpace : public MemorySpace
+{
+public:
+    /// Memory space number for CVs
+    static constexpr uint8_t SPACE_ID =
+        openlcb::MemoryConfigDefs::SPACE_FIRMWARE;
+
+    /// Constructor.
+    /// @param service Service instance to bind this flow to
+    /// @param tx_flow reference to the transmit flow
+    /// @param rx_flow reference to the receive flow
+    FirmwareSpace(
+        Service *service, TxFlowInterface *tx_flow, RxFlowInterface *rx_flow)
+        : MemorySpace(service, tx_flow, rx_flow)
+    {
+    }
+
+private:
+    /// Test if the memory space is read only.
+    /// @return false
+    bool read_only() override
+    {
+        return false;
+    }
+
+    /// Get the largest supported address.
+    /// @return 1023 (max CV address)
+    address_t max_address() override
+    {
+        return UINT32_MAX;
+    };
+
+    /// Get the space ID that will be used over the modem interface.
+    /// @return space id
+    uint8_t get_space_id() override
+    {
+        return SPACE_ID;
+    }
+
+    /// Reboot into bootloader request.
+    /// @return openlcb::Defs::ErrorCodes::ERROR_CODE_OK
+    virtual errorcode_t freeze()
+    {
+        txFlow_->send_packet(Defs::get_reboot_payload(Defs::RebootArg::BOOT));
+        return openlcb::Defs::ErrorCodes::ERROR_CODE_OK;
+    }
+
+    /// Reboot into application with full validation request.
+    /// @return openlcb::Defs::ErrorCodes::ERROR_CODE_OK
```

Comment:
`/// @todo make the unfreeze API abel to suppport asynchronous commands and implement a checksum verification return path`

Reply:
Will address on another pull request.